### DISCUSSION TITLE
Simplify implementing and using custom datastores

### DISCUSSION THREAD

**Discussion by dominikschlosser** [keycloak/keycloak#26077] [Ideas]
Hi,

as already discussed in #25830, the current situation after map-storage removal is a bit challenging, if one wants to run a custom datastore including caching layer.

**Problem**

We identified the following problems:
- `DatastoreProvider` is a good starting point, but sadly not enough, because some default providers register themselves as event listeners or initialize things like infinispan caches in their `init()`-method. This wasn't the case with map-storage, since these providers where disabled when this feature was active (using `EnvironmentDependentProviderFactory`).
- Another related issue is, that JPA- and Infinispan-BuildSteps are always active now. Because of this, we needed to provide a few alternative implementations for `cluster` and `authorizationCache` SPIs.
- Furthermore, we need to extend `LegacyDatastore` because of https://github.com/keycloak/keycloak/discussions/25898
- Finally there are seemingly unrelated SPIs like `globalLock` (which is to be removed to our knowledge), `deploymentState` and `publicKeyStorage` which need to be implemented in addition to `datastore`.

A user of our datastore needs to set these configuration options in order for it to work:

```
- KC_SPI_DATASTORE_LEGACY_ENABLED=false
- KC_SPI_DATASTORE_PROVIDER=cassandra
- KC_SPI_AUTHENTICATION_SESSIONS_INFINISPAN_ENABLED=false
- KC_SPI_USER_SESSIONS_INFINISPAN_ENABLED=false
- KC_SPI_SINGLE_USE_OBJECT_INFINISPAN_ENABLED=false
- KC_SPI_LOGIN_FAILURE_INFINISPAN_ENABLED=false
- KC_SPI_AUTHORIZATION_CACHE_INFINISPAN_ENABLED=false
- KC_SPI_AUTHORIZATION_CACHE_PROVIDER=disabled
- KC_SPI_CLUSTER_INFINISPAN_ENABLED=false
- KC_SPI_CLUSTER_QUARKUS_ENABLED=false
- KC_SPI_CLUSTER_PROVIDER=disabled
- KC_SPI_DEPLOYMENT_STATE_JPA_ENABLED=false
- KC_SPI_DEPLOYMENT_STATE_PROVIDER=cassandra
- KC_SPI_PUBLIC_KEY_STORAGE_INFINISPAN_ENABLED=false
- KC_SPI_PUBLIC_KEY_STORAGE_PROVIDER=cassandra
- KC_SPI_GLOBAL_LOCK_DBLOCK_ENABLED=false
- KC_SPI_GLOBAL_LOCK_PROVIDER=none
- KC_SPI_STICKY_SESSION_ENCODER_INFINISPAN_ENABLED=false
- KC_SPI_STICKY_SESSION_ENCODER_PROVIDER=disabled
- KC_SPI_CONNECTIONS_JPA_LEGACY_ENABLED=false
```

Despite all this configuration, it is currently impossible to disable Infinispan completely (since the corresponding BuildStep is always active). So we have to start Keycloak with `cache=local` to minimize the impact.

`cluster` and `authorizationCache` providers are implemented by us, but do nothing except returning empty lists...

**Proposal**
As discussed in #25830, we aren't happy with the amount of configuration needed to replace Keycloaks storage after map-storage removal. A mechanism like the feature-driven `isSupported()` method in these providers to disable all of them by one external setting would be great. This is why we implemented our own version of this for now, which is only possible by re-using the IDs of the existing providers to prevent them from initializing.

There also needs to be a way to disable JPA and especially Infinispan-BuildSteps like it was with map storage.

And finally i think it would be good to move SPIs like `deploymentState` and `publicKeyStorage` into `DatastoreProvider`, because they are clearly affected by choice of storage implementation. Alternatively there need to be implementations which do not depend on a specific storage technology.

We are willing and able to support you to implement those changes, as i already pointed out.

What do you think?

**Comment by mhajas** [comment]
Thank you for starting this discussion! 

I already created an enhancement request for not running the JPA and Infinispan build steps here: 
- https://github.com/keycloak/keycloak/issues/26054
- https://github.com/keycloak/keycloak/issues/26081

I believe this is a valid requirement and we should not put such obstacles on people trying to implement their storage.

**Comment by xgp** [comment]
:+1: for this proposal

* I would add `EventStoreProvider` to the list of suggestions of SPIs that should be moved into `DatastoreProvider`
* Generally making it easier to override only one SPI is a priority for us, as we often find some implementation detail that we need to fix/enhance, but, as @dominikschlosser mentioned, it takes a ton of configuration to override and stop the build/loading of the defaults.

Also, for specific use cases, we are working on a JPA replacement to the just the parts that use Infinispan (authSession, loginFailure, singleUseObject, userSession), so we support "DatastoreProvider extend LegacyStoreManager so each implementation of the DatastoreProvider needs to provide a way to access each layer of the storage separately" in https://github.com/keycloak/keycloak/discussions/25898
